### PR TITLE
[WIP] MKL repos download for mac and windows + update to MKL(-DNN) 0.12

### PR DESCRIPTION
**Pull request by anton-matosov**
This PR implements MKL repos for Mac (tested) and Windows (not tested yet)
As well as reworks MKL.BUILD to support all 3 desktop platforms

**Update 1** Unfortunately, because of bazel bug https://github.com/bazelbuild/bazel/issues/4480 you will still experience loader issues on Mac and you will be forced to install MKL globally or deal with DYLD_LIBRARY_PATH to overcome invalid rpath in generated .so files of tensorflow (similar to https://github.com/tensorflow/tensorflow/issues/6729 for CUDA libraries in dependencies).

**Update 2** Bazel 0.12.0 (not yet released) fixes the issue and it builds and runs flawlessly (verified with [0.12.0rc1](https://releases.bazel.build/0.12.0/rc1/index.html) )

### CODE REVIEW

File: tensorflow/workspace.bzl [hunk 1]
```diff
@@ -60,6 +60,11 @@ def check_bazel_version_at_least(minimum_bazel_version):
     fail("\nCurrent Bazel version is {}, expected at least {}\n".format(
         native.bazel_version, minimum_bazel_version))
 
+# Sanitize a dependency so that it works correctly from code that includes
+# TensorFlow as a submodule.
+def clean_dep(dep):
```
Comment (gunan) [inline tensorflow/workspace.bzl]: Let's either modify the whole file, or split this out of this PR as a separate PR.
Just want to make sure the file is consistent.

**Review by gunan — CHANGES_REQUESTED** [review]
All build and bzl changes are OK with me except for the one comment. @andydavis1 could you take a look at the cc file changes? I am especially worried about the return type change in `mkl_util.h`, `SetUsrMemHandle`

File: tensorflow/workspace.bzl (hunk 1 above)
Comment (anton-matosov) [reply tensorflow/workspace.bzl]: Sure. I will update all occurrences, if that is acceptable =)

**Comment by anton-matosov** [comment]
@gunan `mkl_util.h` changes fixes a bug, which breaks even linux build. it was discussed here https://github.com/tensorflow/tensorflow/pull/17392#pullrequestreview-101268242

File: tensorflow/workspace.bzl (hunk 1 above)
Comment (anton-matosov) [reply tensorflow/workspace.bzl]: Done

**Comment by anton-matosov** [comment]
Bazel 0.12.0 (not yet released) fixes the linker issue and it builds and runs flawlessly with downloaded MKL libs (verified with 0.12.0rc1 )

**Review by gunan — APPROVED** [review]

**Comment by sb2nov** [comment]
@anton-matosov can you resolve the merge conflicts.

**Comment by anton-matosov** [comment]
@sb2nov done

**Comment by anton-matosov** [comment]
🎉

**Comment by dfumento** [comment]
According to Update 2 Bazel 0.12.0 (not yet released) fixes the issue and it builds and runs flawlessly (verified with 0.12.0rc1 )

I tried with the install version (not homebrew) of 0.12.0 bazel  using this line:

bazel build --config=opt --config=mkl //tensorflow/tools/pip_package:build_pip_package

and I receive this message:
error: unknown target CPU 'armv7-a' 

There are messages on-line that suggest that this is a problem with 0.12.0, but 0.12.0 is needed for the MKL support.

**Comment by anton-matosov** [comment]
bazel 0.12.0 was release couple days ago, check it out (I didn't yet, but should work)
to bypass CPU 'armv7-a' issue with pre-release binary specify `--cpu=darwin` argument  (for mac build) to the bazel call. More details here https://github.com/tensorflow/tensorflow/pull/17392#issuecomment-375420834

**Pull request merged**